### ISSUE TITLE
Operator parsing issue?

### ISSUE THREAD

**Issue by guregu** [ichiban/prolog#219]
Hello, I noticed the following query fails to unify:

```prolog
(A=a+b -> true ; false) = ;(->(=(+(a,b),+(a,b)),true),false).
```

Expected: `A = a+b.`

### Comparison
Running this query:
```prolog
?- A = ;(->(=(+(a,b),+(a,b)),true),false), write_canonical(A), nl.
```

#### ichiban/prolog
```prolog
;(->(=(+(, (a, b)), +(, (a, b))), true), false)
```
Seems like an extra `,` is getting inserted.

#### SWI
```prolog
;(->(=(+(a,b),+(a,b)),true),false)
```

#### Tau
```prolog
';'(->(=(+(a,b),+(a,b)),true),false)

**Comment by ichiban** [comment]
I can replicate the bug with a shorter query:

```go
?- write_canonical(+(a, b)), nl.             
+(, (a, b))
true.
```

Yet, it doesn't look like a bug in the parser:

```go
?- +(a, b) =.. L.               
L = [+, a, b].
```

The bug might be in `func (c *Compound) Unparse(emit func(Token), env *Env, opts ...WriteOption)`. I'll dig into it.

**Comment by guregu** [comment]
My theory is the parser is inserting an extra `','`, and the iterator is flattening it:

### ichiban/prolog
```go
?- +(','(a, b)) =.. L.
L = [+, a, b]

?- +(','(a, b)) = +(a, b).
true
```

### Tau/SWI
```go
?- +(','(a, b)) =.. L.
L = [+,(a,b)].

?- +(','(a, b)) = +(a, b).
false
```

**Comment by ichiban** [comment]
# ichiban/prolgo

# SWI/Scryer

```prolog
?- +(a, b) = +X.
false.
```

**Comment by guregu** [comment]
@ichiban Thanks for tackling this! 💪 

I know it's still a work in progress, but just in case you weren't already aware:
I spotted a regression(?) after the overhaul.

### v0.9.1
```go
?- write_canonical(a: +b).
:(a, +(b))
```
### 5b2fb1b7944d1f96a808a56a8ed8e9b5cd53ea64
```go
?- write_canonical(a: +b).
2022/06/27 18:16:31 failed to query: unexpected token: {letter digit b}
```

Not sure what ISO has to say about this, but Tau and SWI are also `:(a,+(b))`.

**Comment by ichiban** [comment]
@guregu As of newly released `v0.10.0`, the problems are fixed except the one right above.

```console
$ $(go env GOPATH)/bin/1pl
Top level for ichiban/prolog v0.10.0
This is for testing purposes only!
See https://github.com/ichiban/prolog for more details.
Type Ctrl-C or 'halt.' to exit.
?- (A=a+b -> true ; false) = ;(->(=(+(a,b),+(a,b)),true),false).
A = a+b.
?- A = ;(->(=(+(a,b),+(a,b)),true),false), write_canonical(A), nl.
;(->(=(+(a,b),+(a,b)),true),false)
A = a+b=a+b->true;false.
?- +(','(a, b)) =.. L.
L = [+,(a,b)].
?- +(','(a, b)) = +(a, b).
false.
```

The problem with `a: +b` is that we have `op(50, xfx, :)` in `bootstrap.pl` (I don't remember why). Other implementations like swi/scryer have `op(600, xfy, :)` instead. I'll fix it later.

You can work around by overriding `op(600, xfy, :)` for now.

```console
$ $(go env GOPATH)/bin/1pl
Top level for ichiban/prolog v0.10.0
This is for testing purposes only!
See https://github.com/ichiban/prolog for more details.
Type Ctrl-C or 'halt.' to exit.
?- write_canonical(a: +b).
2022/07/16 15:38:12 failed to query: unexpected token: {graphic +}
?- current_op(P, S, :).
P = 50,
S = xfx.
?- op(600, xfy, :).
true.
?- write_canonical(a: +b).
:(a,+(b))true.
```

**Comment by ichiban** [comment]
Fixed!

```console
$ go install github.com/ichiban/prolog/cmd/1pl@latest
go: downloading github.com/ichiban/prolog v0.10.1
$ $(go env GOPATH)/bin/1pl
Top level for ichiban/prolog v0.10.1
This is for testing purposes only!
See https://github.com/ichiban/prolog for more details.
Type Ctrl-C or 'halt.' to exit.
?- current_op(P, S, :).
P = 600,
S = xfy.
?- write_canonical(a: +b).
:(a,+(b))true.
```

**Comment by guregu** [comment]
Thanks for all the fixes!

Found one more:

```go
?- write('[]'), nl.
2022/08/01 04:53:13 failed to query: unexpected token: {end .}
```

**Comment by ichiban** [comment]
Fixed in v0.10.4. `[]` consists of two tokens but `'[]'` consists of one. I messed up pushing back tokens for the latter.
The same goes for `{}` and `'{}'`.

```console
$ $(go env GOPATH)/bin/1pl
Top level for ichiban/prolog v0.10.4
This is for testing purposes only!
See https://github.com/ichiban/prolog for more details.
Type Ctrl-C or 'halt.' to exit.
?- write('[]'), nl.
[]
true.
?- write('{}'), nl.
{}
true.
?- 
```

**Comment by guregu** [comment]
Thanks! Everything looks good now 🎉
